search_public.py

```diff
@@ -120,29 +109,19 @@ def main():
             "startTime": f"{yesterday_formatted}",
             "endTime": f"{now_formatted}"
         }
-        dest_ipHeaders = {
-            "accept": "application/json",
-            "content-type": "application/json",
-            "authorization": "Bearer {}".format(bearer_token)
-        }
-        dest_ipSearchResponse = requests.post(searchUrl, json=dest_ipDataValidationpayload, headers=dest_ipHeaders)
-        print(dest_ipSearchResponse)
-        rawData = json.loads(dest_ipSearchResponse.text) 
-        try:
-            with open(f"{workingDir}/dest_ip_field_output_{now_formatted}.txt", 'w+') as file:
-                fieldNames = ['dest_ip','count','msg_type']
-                writer = csv.DictWriter(file, fieldnames=fieldNames)
-                writer.writeheader()
-                if isinstance(rawData, dict) and 'rows' in rawData:
-                    for data in rawData['rows']:
-                        writer.writerow(data)
-        except OSError as e:
-            print(f"An error occurred: {e}")
-
-
-
-
-
+        tasks.append(("dest_ip_field_output", dest_ip_payload, common_headers, ["dest_ip", "count", "msg_type"]))
+
+    with concurrent.futures.ThreadPoolExecutor(max_workers=4) as executor:
```

Comment:
Why 4? And why a constant hidden in the code ;-)
Also consider using `thread_name_prefix` parameter in this call, set it to something sensible for simple monitoring ...

Reply:
Also, what happens on SIGHUP, docs say one is supposed to do some signal jugglery in the main process:

> This means exceptions in the main thread must be caught and handled in order to signal threads to exit gracefully.

Reply:
ThreadPoolExecutor max_workers constant was set at 4 via laziness. After testing the new logic, that could be set dynamically based on number of CPU cores (e.g. via `multiprocessing` module). 

Great thinking about `thread_name_prefix`. 

The `signal` module gives some control on handling signals like SIGHUP among others. 

I think these would be good items to implement after validating the new threading logic and refactored code flow function as expected.

Reply:
The search script should be rather light weight since it is just hitting the public API and downloading data.  I believe that the bottleneck here would be the speed in which the search platform can offer up results and then the speed of the CPU.  

Even if we increase the thread from 1 to 4 or X, we still have the issue of O(n^2/2X).  In my previous run with 100,000 entries the matrix step was going to take ~400 hours.  Even if we throw all the cores of a laptop at this problem, it will still take an unacceptable amount of time.  I feel that if we can perform some basic normalization and data sanity before the matrix calculations we would be in better shape.